sdk/tables/Azure.Data.Tables/src/TableInternalRestClient.cs

```diff
@@ -354,6 +353,45 @@ internal HttpMessage CreateDeleteEntityRequest(string table, string partitionKey
             return message;
         }
 
+        internal HttpMessage CreateInsertEntityRequest(string table, int? timeout, string requestId, IDictionary<string, object> tableEntityProperties, QueryOptions queryOptions)
```

Comment:
Wonder if we can do something in generator to allow you to redefine less code. Maybe a custom converter annotations for properties?

Reply:
This is all just so that we serialize long as string, which Table requires :(

Reply:
Wonder if it's worth to remapping in the public client? We can also add/remove ODAta annotations there.

Reply:
What's your perspective on the benefits of doing that?

Reply:
Yeah, the less generated code we take over the easier it would be both to update generator and react to swagger updates.

It's also easier to follow if customizations are in one place instead of spread along.

The last point is that this code might be simpler if it would operate on Dictionary<string, object> rather than on JSON.

Reply:
My only concern with not doing in inline during serialization is that I think it will be less efficient. But your points are valid.

Have we considered generating code in a way that allows for more discrete customizations for serializations? For example: 
```csharp
internal HttpMessage CreateInsertEntityRequest(string table, int? timeout, string requestId, IDictionary<string, object> tableEntityProperties, QueryOptions queryOptions)
{
    var message = _pipeline.CreateMessage();
    var request = message.Request;
    request.Method = RequestMethod.Post;
    var uri = new RawRequestUriBuilder();
    uri.AppendRaw(url, false);
    uri.AppendPath("/", false);
    uri.AppendPath(table, true);
    if (timeout != null)
    {
        uri.AppendQuery("timeout", timeout.Value, true);
    }
    if (queryOptions?.Format != null)
    {
        uri.AppendQuery("$format", queryOptions.Format.Value.ToString(), true);
    }
    request.Uri = uri;
    request.Headers.Add("x-ms-version", version);
    if (requestId != null)
    {
        request.Headers.Add("x-ms-client-request-id", requestId);
    }
    request.Headers.Add("DataServiceVersion", "3.0");
    request.Headers.Add("Content-Type", "application/json;odata=nometadata");
    request.Content = InsertEntitySerializer.CreateContent(tableEntityProperties);
    return message;
}
```
InsertEntitySerializer.cs
```csharp
RequestContent Create(IDictionary<string, object> tableEntityProperties)
{
	if (tableEntityProperties != null)
	{
	    using var content = new Utf8JsonRequestContent();
	    content.JsonWriter.WriteStartObject();
	    foreach (var item in tableEntityProperties)
	    {
	        SerializeEntityPropertyItem(content, item);
	    }
	    content.JsonWriter.WriteEndObject();
	    request.Content = content;
	}
}
```

Reply:
Why do you think it would be less efficient?

> Have we considered generating code in a way that allows for more discrete customizations for serializations? For example:

There were ideas floating about allowing to assign converters to model properties that generated code would call into but we didn't consider anything for method parameters yet.

Reply:
> Why do you think it would be less efficient?

Thinking more about it, the cost may be pretty close to the same. The remaining concern would be that we'd be modifying the object provided by the caller on serialize, as opposed to only augmenting what goes over the wire. On de-serialization, it's almost the same.

Reply:
From the perspective of someone without much familiarity with how our generation/override norms, Chris' suggestion makes a lot of sense to me.  I can see the desire to take ownership of forming the content body for requests as being something useful and not uncommon.   Having the generator break it out as a discrete step that could be customized seems that it would be a good way to reduce the surface area of overrides.

The other thing that occurs to me would be to apply an AOP pattern for Before/After interceptors, so that customization could massage parameters being passed to the generated method (the Before) or modify the output of the generated method (the After) - though in this kind of scenario it wouldn't be any more helpful than moving to the public class;  you'd still have to pay the cost of creating the content twice.

Reply:
We definitely shouldn't modify customers object, rather making a copy and adding new properties there.

On deserialization code-path we can both modify or copy.

Reply:
Agreed. Then if we copy, that would incur an avoidable allocation.

Reply:
I think at this stage of the project we might prefer maintainability of the code base over perf especially if it's about a single additional allocation.

Reply:
It would be one allocation per insert, which could get expensive in bulk insert scenarios. 

I think the other consideration is whether or not we'll have to do this anyway for developer defined strongly typed entity scenarios.

Reply:
>It would be one allocation per insert,

Not if you reuse the dictionary :)

Reply:
> > It would be one allocation per insert,
> 
> Not if you reuse the dictionary :)

That would get interesting if someone writes a parallel Insert statement :)

Reply:
I was hoping we'd be able to do some of this with partial methods, but I agree with the general principal that we should do whatever we can to prevent people from copy/pasting the entire protocol method when they want to change one small part.  A couple of extra allocations when customizing would be worth it to me.

Reply:
Thanks for the additional perspective. I’ll refactor this into the public’s client as suggested.